File: dotnet/Capcom/GameX.Capcom/Formats/PakBinary_Arc.cs
using GameX.Formats;
using System;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using static GameX.Util;

namespace GameX.Capcom.Formats
{
    public unsafe class PakBinary_Arc : PakBinary<PakBinary_Arc>
    {
        // Header
        #region Header

        const uint K_MAGIC = 0x00435241;

        [StructLayout(LayoutKind.Sequential, Pack = 0x1)]
        struct K_Header
        {
            public ushort Version;
            public ushort NumFiles;
        }

        [StructLayout(LayoutKind.Sequential, Pack = 0x1)]
        struct K_File
        {
            public fixed byte Path[0x40];
            public uint Compressed;
            public uint PackedSize;
            public uint FileSize;
            public uint Offset;
        }

        #endregion

        public override Task Read(BinaryPakFile source, BinaryReader r, object tag)
        {
            var magic = r.ReadUInt32();
            magic &= 0x00FFFFFF;
            if (magic != K_MAGIC) throw new FormatException("BAD MAGIC");

            // get header
            var header = r.ReadT<K_Header>(sizeof(K_Header));

            // get files
            source.Files = r.ReadTArray<K_File>(sizeof(K_File), header.NumFiles)
                .Select(x => new FileSource
                {
                    Path = $"{Encoding.UTF8.GetString(new Span<byte>(x.Path, 0x40)).TrimEnd('\0')}{GetExtension(r, x.Offset)}".Replace('\\', '/'),
                    Compressed = (int)x.Compressed,
                    PackedSize = x.PackedSize,
                    FileSize = x.FileSize,
                    Offset = x.Offset,
                }).ToArray();
            return Task.CompletedTask;
        }

        public override Task<Stream> ReadData(BinaryPakFile source, BinaryReader r, FileSource file, FileOption option = default)
        {
            r.Seek(file.Offset);
            return Task.FromResult<Stream>(new MemoryStream(Decompress(r, (int)file.PackedSize, (int)file.FileSize)));
        }

        static byte[] Decompress(BinaryReader r, int length, int newLength = 0) => r.DecompressZlib(length, newLength);

        static string GetExtension(BinaryReader r, long position)
        {
            r.Seek(position);
            return _guessExtension(Decompress(r, 150));
        }
    }
}

File: Python/gamex/Capcom/pakbinary_arc.py
import os
from io import BytesIO
from gamex.filesrc import FileSource
from gamex.pak import PakBinaryT
from gamex.compression import decompressZlib, decompressZstd
from gamex.util import _guessExtension

# typedefs
class Reader: pass
class BinaryPakFile: pass

# PakBinary_Arc
class PakBinary_Arc(PakBinaryT):

    #region K

    K_MAGIC = 0x00435241
    class K_Header:
        struct = ('<HH', 4)
        def __init__(self, tuple):
            self.version, \
            self.numFiles = tuple
    class K_File:
        struct = ('<64sIIII', 80)
        def __init__(self, tuple):
            self.path, \
            self.compress, \
            self.zsize, \
            self.size, \
            self.offset = tuple

    #endregion

    # read
    def read(self, source: BinaryPakFile, r: Reader, tag: object = None) -> None:
        magic = r.readUInt32()
        if magic != self.K_MAGIC: raise Exception('BAD MAGIC')

        header = r.readS(self.K_Header)
        kfiles = r.readTArray(self.K_File, header.numFiles)

        # get files
        source.files = files = [FileSource(
            compressed = x.compress,
            path = x.path.decode('utf8').strip('\x00').replace('\\', '/'),
            packedSize = x.zsize,
            fileSize = x.size,
            offset = x.offset
        ) for x in kfiles]

        # add extension
        for file in files:
            r.seek(file.offset)
            buf = _decompress(r, 150, full=False)
            file.path += _guessExtension(buf)

    # readData
    def readData(self, source: BinaryPakFile, r: Reader, file: FileSource) -> BytesIO:
        r.seek(file.offset)
        return BytesIO(_decompress(r, file.compressed, file.packedSize, file.fileSize))

@staticmethod
def _decompress(r: Reader, length: int, newLength: int=0, full: bool=True) -> bytes:
    return decompressZlib(r, length, newLength, full=full)